### DISCUSSION TITLE
Question about uefi capsule integration into swupdate

### DISCUSSION THREAD

**Discussion by tschoop** [OE4T/meta-tegra#1592] [General]
I have got a working firmware update solution using kirkstone/JP35.5 using swupdate with A/B. works nice.
Now I want to integrate the low-level bootloader stuff into the swupdate.
I managed to integrate the uefi capsule into the swupdate package and update process.

But now I am not sure if this is safe in regards to powerfailure and stability. So my question is: is there a possibility to readout any version information of the capsule image and the low level bootloaders to check whether an update of the uefi capsule is needed? 
If I had such a version check, I  could package and deliver the current uefi capsule in all swupdate packages and could skip the real flashing if not needed.
Any hints appreciated.

**Comment by madisongh** [comment]
There is a version number put into the generated capsule, based on the BSP version.  That wouldn't catch changes you've made that might affect the capsule contents, though (e.g., patching UEFI or updating your device tree).

The capsule update should handle power failures or other interruptions - the update gets applied to the unused boot slot, then a switchover happens once that's complete. Of course, it's always best to test that to be sure.

**Reply by tschoop** [reply]
Thanks. I will give it a try. If I cannot distinguish between the capsules, I will do it via a checksum.
